awx/main/scheduler/task_manager.py

```diff
@@ -267,185 +256,47 @@ class DependencyManager(TaskBase):
-            if "inventory_sources_already_updated" in start_args and inventory_source.id in start_args['inventory_sources_already_updated']:
-                continue
-            if not inventory_source.update_on_launch:
-                continue
-            latest_inventory_update = self.get_latest_inventory_update(inventory_source)
-            if self.should_update_inventory_source(task, latest_inventory_update):
-                inventory_task = self.create_inventory_update(task, inventory_source)
-                created_dependencies.append(inventory_task)
-                dependencies.append(inventory_task)
-            else:
-                dependencies.append(latest_inventory_update)
-
-        if dependencies:
-            self.add_dependencies(task, dependencies)
-
-        return created_dependencies
-
-    def gen_dep_for_inventory_update(self, inventory_task):
-        created_dependencies = []
-        if inventory_task.source == "scm":
-            invsrc = inventory_task.inventory_source
-            if not invsrc.source_project.scm_update_on_launch:
-                return created_dependencies
-
-            latest_src_project_update = self.get_latest_project_update(invsrc.source_project_id)
-            if self.should_update_related_project(inventory_task, latest_src_project_update):
-                latest_src_project_update = self.create_project_update(inventory_task, project_id=invsrc.source_project_id)
-                created_dependencies.append(latest_src_project_update)
-            self.add_dependencies(inventory_task, [latest_src_project_update])
-            latest_src_project_update.scm_inventory_updates.add(inventory_task)
-        return created_dependencies
-
-    @timeit
-    def generate_dependencies(self, undeped_tasks):
-        created_dependencies = []
-        for task in undeped_tasks:
-            task.log_lifecycle("acknowledged")
-            if type(task) is Job:
-                created_dependencies += self.gen_dep_for_job(task)
-            elif type(task) is InventoryUpdate:
-                created_dependencies += self.gen_dep_for_inventory_update(task)
-            else:
-                continue
-        UnifiedJob.objects.filter(pk__in=[task.pk for task in undeped_tasks]).update(dependencies_processed=True)
-
-        return created_dependencies
-
-    def process_tasks(self):
-        deps = self.generate_dependencies(self.all_tasks)
-        self.generate_dependencies(deps)
-        self.subsystem_metrics.inc(f"{self.prefix}_pending_processed", len(self.all_tasks) + len(deps))
-
     @timeit
     def _schedule(self):
-        self.get_tasks(dict(status__in=["pending"], dependencies_processed=False))
-
-        if len(self.all_tasks) > 0:
-            self.get_inventory_source_tasks()
-            self.process_tasks()
+        failing_job_qs = UnifiedJob.objects.filter(
+            status='pending', dependencies_processed=False, dependent_jobs__status__in=['failed', 'error']
```

Comment:
I want to call attention to the "canceled" status as a weird gap in queryset completeness here.

Here, we fail jobs with dependencies in the failed/error status. In the block after this, we advance pending jobs that don't have any dependencies in the failed/error/canceled status. What about canceled dependencies? This is saying that we don't fail them, and we don't advance them either.

That's right. Here, the `DependencyManager` doesn't do _anything_ with jobs that had a dependency get canceled. This is actually pre-existing behavior. It was moved and some lower-level details changed, but it was always the case that the task manager didn't do anything when dependencies got canceled. Why? Because that was/is handled in the `cancel` method, we propagate the cancel from the dependency to jobs that depend on it. That's what the (new name) `get_cancel_chain` loop does.

Would it be _easier_ to not propagate through the "hard" dependency chain in the `cancel` method, and do it here instead? Yes, I think it would be. It would be easier to follow (because this manages dependencies, seems only right to manage canceling through deps).

My only reason not to is to have a stopping place. I'm still maybe, kind of, considering it.

Reply:
Never mind, I've given up on this because it won't happen in a very satisfying way.